src/EFCore.PG/Storage/Internal/NpgsqlDataSourceManager.cs

```diff
@@ -70,39 +68,39 @@ public NpgsqlDataSourceManager(IEnumerable<INpgsqlDataSourceConfigurationPlugin>
             { ConnectionString: null } or null => null,
 
             // The following are features which require an NpgsqlDataSource, since they require configuration on NpgsqlDataSourceBuilder.
-            { EnumDefinitions.Count: > 0 } => GetSingletonDataSource(npgsqlOptionsExtension, "MapEnum"),
-            _ when _plugins.Any() => GetSingletonDataSource(npgsqlOptionsExtension, _plugins.First().GetType().Name),
+            { EnumDefinitions.Count: > 0 } => GetSingletonDataSource(npgsqlOptionsExtension),
+            _ when _plugins.Any() => GetSingletonDataSource(npgsqlOptionsExtension),
 
             // If there's no configured feature which requires us to use a data source internally, don't use one; this causes
             // NpgsqlRelationalConnection to use the connection string as before (no data source), allowing switching connection strings
             // with the same service provider etc.
             _ => null
         };
 
-    private DbDataSource GetSingletonDataSource(NpgsqlOptionsExtension npgsqlOptionsExtension, string dataSourceFeature)
+    private DbDataSource GetSingletonDataSource(NpgsqlOptionsExtension npgsqlOptionsExtension)
     {
-        if (!_isInitialized)
+        var connectionString = npgsqlOptionsExtension.ConnectionString;
+        Check.DebugAssert(connectionString is not null, "Connection string can't be null");
+
+        if (_dataSources.TryGetValue(connectionString, out var dataSource))
         {
-            lock (_lock)
-            {
-                if (!_isInitialized)
-                {
-                    _dataSource = CreateSingletonDataSource(npgsqlOptionsExtension);
-                    _connectionString = npgsqlOptionsExtension.ConnectionString;
-                    _isInitialized = true;
-                    return _dataSource;
-                }
-            }
+            return dataSource;
         }
 
-        Check.DebugAssert(_dataSource is not null, "_dataSource cannot be null at this point");
+        var newDataSource = CreateDataSource(npgsqlOptionsExtension);
 
-        if (_connectionString != npgsqlOptionsExtension.ConnectionString)
+        var addedDataSource = _dataSources.GetOrAdd(connectionString, newDataSource);
+        if (!ReferenceEquals(addedDataSource, newDataSource))
         {
-            throw new InvalidOperationException(NpgsqlStrings.DataSourceWithMultipleConnectionStrings(dataSourceFeature));
+            newDataSource.Dispose();
+        }
+        else if (_isDisposed == 1)
```

Comment:
Not sure what the infra looks like - in terms of full disposal blocking this method from being called later on - but you might want this check to be done before as well. Just so you don't keep adding more items to the dictionary when things are already disposed.

Reply:
> Not sure what the infra looks like

Me neither... My logic here is that it's OK if a data source gets added to the dictionary during/after Dispose (after all Dispose doesn't clear the dictionary, we don't really care) - the important bit is that the data sources always get disposed, which I think this achieves...

Will go ahead and merge, we can always revisit if needed (unlikely).